Add pair style lj/pirani

mateors6 opened the pull request:
**Summary**

This pull adds a new pair style: an Improved Lennard-Jones (ILJ) potential, by Pirani et al. Phys. Chem. Chem. Phys., 2008,10, 5489-5503


**Related Issue(s)**


**Author(s)**

Mateo Rodríguez (mateorsuarez@gmail.com) ( Molecular Interactions Group INTERMOL IFF-CSIC, http://intermol.iff.csic.es)


**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**


**Implementation Notes**


**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

mateors6 committed:
Add files via upload

mateors6 committed:
Add files via upload

mateors6 committed:
Delete doc/pair_ilj_cut.rst

mateors6 committed:
Add files via upload

mateors6 committed:
Merge branch 'develop' into ILJ/CUT

mateors6 committed:
Delete doc/pair_ilj_cut.html

mateors6 committed:
Update Commands_pair.rst

mateors6 committed:
Update pair_style.rst

mateors6 committed:
Merge branch 'lammps:develop' into lj/improved/cut

mateors6 committed:
Update .gitignore

mateors6 committed:
Delete src/EXTRA-PAIR/pair_ilj_cut.cpp

mateors6 committed:
Delete src/EXTRA-PAIR/pair_ilj_cut.h

mateors6 committed:
Add files via upload

mateors6 committed:
Update and rename pair_ilj_cut.rst to pair_lj_improved_cut.rst

mateors6 committed:
Update pair_lj_improved_cut.rst

mateors6 committed:
Update .gitignore

akohlmey commented on doc/src/pair_lj_improved_cut.rst:
```diff
@@ -0,0 +1,153 @@
+.. index:: pair_style lj/improved/cut
+
+pair_style lj/improved/cut command
+=============================
```
```suggestion
==================================
```
Underlines must match the length of the text or sphinx is unhappy.

akohlmey committed:
add force style test

akohlmey committed:
replace slow pow(x,2.0) with square(x) and cache repeatedly used pow() calls

akohlmey committed:
address multiple issues with formatting, references, non-ASCII characters, spelling

akohlmey commented:
@mateors6 I've added some simple optimizations (you want to check with some existing input how much of a difference that makes), fixed a few issues with the doc file, and added a test input.
That test produces invalid forces for multi-cutoff r-RESPA, so I have (temporarily) disabled it.  Please test carefully and either correct the issue or remove the code. Single-cutoff r-RESPA passes.

mateors6 committed:
Update pair_lj_improved_cut.cpp

mateors6 commented:
Dear @akohlmey,

Thank you. With your optimizations, now it runs around 1% faster than before (with a little run, I'll see how it works with a longer one). 
I have updated the .cpp file, now both single-cutoff and multi-cutoff r-RESPA should be working. I'm going to check if I have updated the file correctly. 

Thanks for your help.

akohlmey commented:
> I have updated the .cpp file, now both single-cutoff and multi-cutoff r-RESPA should be working

Forces seem OK, but energy not.

mateors6 commented:
> > I have updated the .cpp file, now both single-cutoff and multi-cutoff r-RESPA should be working
> 
> Forces seem OK, but energy not.

Yes, I am checking it. I will update once it's solved.

akohlmey committed:
add another small optimization

akohlmey commented:
@mateors6 here is another simple optimization (reduce pointer chasing when updating forces).

mateors6 committed:
Update pair_lj_improved_cut.cpp

Corrected energy calculation for r-RESPA

mateors6 commented:
> @mateors6 here is another simple optimization (reduce pointer chasing when updating forces).

That's interesting! I didn't know it was better that way. 

If I have uploaded the .cpp file correctly, now the energy calculation with r-RESPA should be OK.

akohlmey commented:
> That's interesting! I didn't know it was better that way.

There are 3 simple changes that can be applied to most pair styles that can improve performance.
- cache outer loop forces in local variables (which I just applied)
- change from array of arrays to array of structs for positions and forces
- move if statements (on whether to compute and tally energy and stress) out of the inner loops by templating the compute function
You can see these in many pair styles of the OPT and OPENMP packages.

mateors6 commented:
> > That's interesting! I didn't know it was better that way.
> 
> There are 3 simple changes that can be applied to most pair styles that can improve performance.
> 
> * cache outer loop forces in local variables (which I just applied)
> * change from array of arrays to array of structs for positions and forces
> * move if statements (on whether to compute and tally energy and stress) out of the inner loops by templating the compute function
>   You can see these in many pair styles of the OPT and OPENMP packages.

That's great, thank you. I'll take a look at it.

sjplimp reviewed (changes requested):
I would like to look this over next week to see if I have comments

sjplimp commented:
Here is my first (perhaps only) comment - I don't think the name lj/improved is sufficiently descriptive.
"Improved" can mean a number of things in the context of an empirical potential.
If you look at all the other pair lj variants in LAMMPS, they all have more descriptive names than "improved".
And their authors probably all think their potential is "improved" in some respect for some use cases vs pure LJ.
So what would be a more descriptive name?  Possibly lj/pirani.  You also don't need to have "cut" in the name
unless you are planning a lj/something/long variant in the future.  Did they give this potential a name
in the original paper - it they simply called it "improved LJ", I'm still thinking "no" for LAMMPS.

mateors6 commented:
> Here is my first (perhaps only) comment - I don't think the name lj/improved is sufficiently descriptive.
> "Improved" can mean a number of things in the context of an empirical potential.
> If you look at all the other pair lj variants in LAMMPS, they all have more descriptive names than "improved".
> And their authors probably all think their potential is "improved" in some respect for some use cases vs pure LJ.
> So what would be a more descriptive name?  Possibly lj/pirani.  You also don't need to have "cut" in the name
> unless you are planning a lj/something/long variant in the future.  Did they give this potential a name
> in the original paper - it they simply called it "improved LJ", I'm still thinking "no" for LAMMPS.

It has been named 'Improved Lennard-Jones' since it has more parameters than the LJ and the LJ can be recovered from this one. So, in the end, it's just a more complex approach. Also, it "eliminates most of the inadequacies at short- and long-range of the LJ model.", as stated in the original paper, where it is named 'ILJ'. 
Probably lj/pirani is a suitable name. We are not planning on adding new variants.
I'll ask my supervisors to see what they think about this topic.
Thanks for your comment.

mateors6 committed:
Update Commands_pair.rst

New name

mateors6 committed:
Update and rename pair_lj_improved_cut.rst to pair_lj_pirani.rst

New name

mateors6 committed:
Update pair_style.rst

New name

mateors6 committed:
Update .gitignore

New name

mateors6 committed:
Update and rename pair_lj_improved_cut.cpp to pair_lj_pirani.cpp

New name

mateors6 committed:
Update and rename pair_lj_improved_cut.h to pair_lj_pirani.h

New name

mateors6 committed:
Update and rename mol-pair-lj_improved_cut.yaml to mol-pair-lj_pirani.yaml

New name
